### PR TITLE
Fix jsonpb to serialize NaN and Infinity as special strings

### PR DESCRIPTION
**Pull request by narqo**
As described in #231 float and double fields, `nan` and `(-)inf` must be serialized to strings `"NaN"`, `"Infinity"`, and `"-Infinity"`.

### CODE REVIEW

**Comment by narqo** [comment]
I signed it!

File: jsonpb/jsonpb_test_proto/test_objects.proto [hunk 1]
```diff
@@ -54,6 +54,13 @@ message Simple {
   optional bytes o_bytes = 11;
 }
 
+// Test message for holding special primitives.
+message Specials {
```
Comment (cybrcodr) [inline jsonpb/jsonpb_test_proto/test_objects.proto]: Let's name this NonFinites or NonFiniteFloats to be more specific.

File: jsonpb/jsonpb_test_proto/test_objects.proto [hunk 2]
```diff
@@ -54,6 +54,13 @@ message Simple {
...
+    optional float o_nan = 1;
+    optional float o_pinf = 2;
+    optional float o_ninf = 3;
```
Comment (cybrcodr) [inline jsonpb/jsonpb_test_proto/test_objects.proto]: Just for completeness, can you also add double fields?

File: jsonpb/jsonpb.go [hunk 3]
```diff
@@ -494,6 +494,24 @@ func (m *Marshaler) marshalValue(out *errWriter, prop *proto.Properties, v refle
 		return out.err
 	}
 
+	// Handle special cases for float, double.
```
Comment (cybrcodr) [inline jsonpb/jsonpb.go]: Sorry, another nit.  Please update comment to be more specific, perhaps ...

// Handle non-finite floats like NaN, Infinity and -Infinity.

**Comment by cybrcodr** [comment]
@narqo I'd like to follow up on this PR.  If I don't hear back next week, I'll submit similar fix in instead.

**Comment by narqo** [comment]
Hey, @cybrcodr I've done a rebase onto master's head and fixed what you'd asked. Hope things look better now.

**Comment by kesutton** [comment]
What about the unmarshaling side?  With this change, jsonpb can not successfully deserialize its own output, as it falls back to the standard json decoder due to error: `json: cannot unmarshal string into Go value of type float64`.

I think something very similar to how string encoded integers are handled here can be done: https://github.com/golang/protobuf/blob/master/jsonpb/jsonpb.go#L969

I wrote up something quick locally right after the linked section.

```
import math
...
// floats can be encoded as strings, like "NaN", "+Infinity", and "-Infinity".  In this case, we
// deserialize into the proper float64 type.  If we don't have a
// string value, we just let the standard unmarshaler parse the float
if targetType.Kind() == reflect.Float32 || targetType.Kind() == reflect.Float64 &&
	strings.HasPrefix(string(inputValue), `"`) {
	s := string(inputValue)

	if s == "\"NaN\"" {
	target.SetFloat(math.NaN())
		return nil
	}
	if s == "\"Infinity\"" {
		target.SetFloat(math.Inf(0))
		return nil
	}
	if s == "\"-Infinity\"" {
		target.SetFloat(math.Inf(-1))
		return nil
	}

	return fmt.Errorf("unknown string value %s for float", s)
}
```

I'm not sure if the check for Float32 is necessary, nor am I sure the decision to return that error message is sound.  Just posting something to get the ball rolling.

**Comment by cybrcodr** [comment]
Sorry, I should have mentioned.  Fix for deserialization is coming soon.

**Comment by cybrcodr** [comment]
Fyi.  Fix for deserialization in 5a0f697c9ed9d68fef0116532c6e05cfeae00e55.

**Pull request merged**